3scale/porta
# THREESCALE-4145: Unify CDN backends

**Which issue(s) this PR fixes** 

[THREESCALE-4145](https://issues.redhat.com/browse/THREESCALE-4145)

**Background**

We provide a liquid tag `cdn_asset` which accepts a limited [list of values](https://github.com/jlledom/porta/blob/THREESCALE-4145-unify-cdn/lib/developer_portal/lib/liquid/tags/cdn_asset.rb#L19-L46). Those refer to assets frequently used from developer portals, mainly jQuery, SwaggerUI and FontAwesome libs.

Thanks to this, the buyers can add something like this to their CMS templates:

`{% cdn_asset /jquery/3.5.0/jquery.min.js %}`

That will set jQuery 3.5.0 ready to use in the developer portal. We make the assets available for buyers just by uploading them to [`/cdn`](https://github.com/3scale/porta/tree/6017b4a916df3b027601a4c455124133b2b4acda/cdn). This allows us to update the assets without requiring providers to update their templates .

This feature is called CDN but it's not really a CDN, since the assets are served from our servers. Besides, this is only available from the developer portal, and completely ignores the Rails standard way to configure a CDN by the `asset_host` config parameter (`RAILS_ASSET_HOST` for us).

However, another environment variable is available: `ASSETS_CDN_HOST`. When set, the `cdn_asset` tag will try to download the assets from the `ASSETS_CDN_HOST` instead of our server... but, that can't be a true CDN, because in practice, a CDN receiving a request for `"#{ASSETS_CDN_HOST}/jquery/3.5.0/jquery.min.js"` will try to fetch `"/jquery/3.5.0/jquery.min.js"` from our servers, but those assets won't be there because setting `ASSETS_CDN_HOST` prevents rails routes to be created. In fact, even if they were created, the routes would look like `"/_cdn_assets_/jquery/3.5.0/jquery.min.js"` and the CDN would fail to fetch the assets due to the `/_cdn_assets_` prefix. Summarizing, the only thing you can do with `ASSETS_CDN_HOST` is to copy the assets somewhere else and serve them from there.

All this logic is implemented in a simple way, at [cdn_asset.rb](https://github.com/3scale/porta/blob/013be400daa96cfef89cc5f8d5cdd28b57f6403a/lib/developer_portal/lib/liquid/tags/cdn_asset.rb#L51-L54). For a path like `/jquery/3.5.0/jquery.min.js`, if `ASSETS_CDN_HOST` is set, it renders `"#{ASSETS_CDN_HOST}/jquery/3.5.0/jquery.min.js"`, otherwise it renders `"/_cdn_assets_/jquery/3.5.0/jquery.min.js"`. All requests sent to `/_cdn_assets_` are [redirected](https://github.com/3scale/porta/blob/013be400daa96cfef89cc5f8d5cdd28b57f6403a/config/routes.rb#L28) to a new [Sprockets environment](https://github.com/3scale/porta/blob/013be400daa96cfef89cc5f8d5cdd28b57f6403a/config/routes.rb#L4-L13) pointing to the `/cdn` folder in our root.

**What this PR does**

The issue asks for 3 things, as I understood it:

1. Remove the additional Sprockets environment and user the Rails default one
2. Serve all assets in porta, including the assets under `/cdn`, using the same CDN.
3. Move all `/cdn` assets to a new directory `/dev-portal`.

This PR implements the 3 points above. They are implemented this way:

1. All assets under `/cdn` have been moved to `/public/dev-portal`. This way the default Sprockets environment will find this assets as statics under the `/dev-portal` path. Since now the path is always the same, it can be reached from our server as well as from a CDN.
2. The additional Sprockets environment and the additional routes have been deleted from `routes.rb`. They are not needed anymore.
3. The `cdn_asset` liquid tag now always renders the `/dev-portal` path, no matter whether a CDN is configured or not. When a CDN is configured, it renders an absolute URL instead of a relative URL.
4. The `ASSETS_CDN_HOST` env variable and the `assets_cdn_host` Rails config have been deleted. Now only one way to set a CDN is allowed: by using `RAILS_ASSET_HOST`.
5. I updated the tests to check the new behavior.

**How to review this PR**

- There are more than 50 changed files, but most of them are the assets being moved to another directory. Those files don't need to be reviewed.
- The point `2` above is implemented in `routes.rb`
- The point `3` above is implemented in `cdn_asset.rb`
- The point `4` above is implemented in the files `settings.yml`

**Breaking changes**

- Since `ASSETS_CDN_HOST` has been removed, anybody using it will have to migrate to `RAILS_ASSET_HOST`.
- The old `ASSETS_CDN_HOST` variable expected a full url with the protocol included: `http[s]://cdn.domain.net`. But `RAILS_ASSET_HOST` only accepts the domain: `cdn.domain.net`. For the later, the protocol is [guessed](https://github.com/3scale/porta/blob/013be400daa96cfef89cc5f8d5cdd28b57f6403a/app/helpers/application_helper.rb#L378) from the received HTTP request, to avoid Mixed Content. Maybe it's not the best thing to do, because the system admin must ensure the CDN server will accept the same protocol as porta, but so far nobody complained. For the `cdn_asset` tag, it does exactly the same now.

**Verification steps** 

1. Find a template where `cdn_asset` is being used. E.g: In a default developer portal, the layout `main_layout` includes `{% cdn_asset /jquery/3.5.0/jquery.min.js %}`
2. Restart the rails server with `RAILS_ASSET_HOST` unset.
3. Enter the developer portal, it should load jquery from `http://provider.3scale.localhost:3000/dev-portal/jquery/3.5.0/jquery.min.js`
4. Restart the rails server with `RAILS_ASSET_HOST=cdn.3scale.localhost:3000`
6. Enter the developer portal, it should load jquery from `http://cdn.3scale.localhost:3000/dev-portal/jquery/3.5.0/jquery.min.js`
